flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/ThreadInfoSampleService.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.flink.runtime.taskexecutor;
+
+import org.apache.flink.api.common.time.Time;
+import org.apache.flink.runtime.concurrent.ScheduledExecutor;
+import org.apache.flink.runtime.messages.ThreadInfoSample;
+import org.apache.flink.runtime.util.JvmUtils;
+import org.apache.flink.runtime.webmonitor.threadinfo.ThreadInfoSamplesRequest;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+import static java.util.Objects.requireNonNull;
+import static org.apache.flink.util.Preconditions.checkNotNull;
+
+/** Samples thread infos of tasks. */
+class ThreadInfoSampleService {
+
+    private final ScheduledExecutor scheduledExecutor;
+
+    ThreadInfoSampleService(final ScheduledExecutor scheduledExecutor) {
+        this.scheduledExecutor =
+                requireNonNull(scheduledExecutor, "scheduledExecutor must not be null");
+    }
+
+    /**
+     * Returns a future that completes with a given number of thread info samples of a task thread.
+     *
+     * @param task The task to be sampled from.
+     * @param requestParams Parameters of the sampling request.
+     * @return A future containing the stack trace samples.
+     */
+    public CompletableFuture<List<ThreadInfoSample>> requestThreadInfoSamples(
+            final ThreadInfoSampleableTask task, final ThreadInfoSamplesRequest requestParams) {
+        checkNotNull(task, "task must not be null");
+
+        return requestThreadInfoSamples(
+                task,
+                requestParams.getNumSamples(),
+                requestParams.getDelayBetweenSamples(),
+                requestParams.getMaxStackTraceDepth(),
+                new ArrayList<>(requestParams.getNumSamples()),
+                new CompletableFuture<>());
```

Comment:
seems a bit odd to have this as a parameter

Reply:
This method is called recursively, hundreds of times per subtask, so initing it within the method body seems wasteful. What would you propose?

Reply:
I removed the return type and made it apparent that this is an `out` parameter.